# Challenge 1 - Tic Tac Toe

In this lab you will perform deep learning analysis on a dataset of playing [Tic Tac Toe](https://en.wikipedia.org/wiki/Tic-tac-toe).

There are 9 grids in Tic Tac Toe that are coded as the following picture shows:

![Tic Tac Toe Grids](tttboard.jpg)

In the first 9 columns of the dataset you can find which marks (`x` or `o`) exist in the grids. If there is no mark in a certain grid, it is labeled as `b`. The last column is `class` which tells you whether Player X (who always moves first in Tic Tac Toe) wins in this configuration. Note that when `class` has the value `False`, it means either Player O wins the game or it ends up as a draw.

Follow the steps suggested below to conduct a neural network analysis using Tensorflow and Keras. You will build a deep learning model to predict whether Player X wins the game or not.

## Step 1: Data Engineering

This dataset is almost in the ready-to-use state so you do not need to worry about missing values and so on. Still, some simple data engineering is needed.

1. Read `tic-tac-toe.csv` into a dataframe.
1. Inspect the dataset. Determine if the dataset is reliable by eyeballing the data.
1. Convert the categorical values to numeric in all columns.
1. Separate the inputs and output.
1. Normalize the input data.

In [1]:
# your code here
# your code here
import pandas as pd
from sklearn.preprocessing import LabelEncoder, StandardScaler

# Step 1: Read the dataset
# Assuming the file is named 'tic-tac-toe.csv' and in the same directory
data = pd.read_csv('tic-tac-toe.csv')

# Step 2: Inspect the dataset
print("First 5 rows of the dataset:")
print(data.head())

print("\nDataset Information:")
print(data.info())

print("\nUnique values in each column:")
for col in data.columns:
    print(f"{col}: {data[col].unique()}")

# Step 3: Convert categorical values to numeric
label_encoder = LabelEncoder()

for col in data.columns[:-1]:  # Skip the 'class' column for now
    data[col] = label_encoder.fit_transform(data[col])

# Encode the target column ('class')
data['class'] = label_encoder.fit_transform(data['class'])

# Step 4: Separate inputs and output
X = data.drop(columns=['class'])  # Features
y = data['class']  # Target

# Step 5: Normalize the input data
scaler = StandardScaler()
X_normalized = scaler.fit_transform(X)

print("\nNormalized Features (first 5 rows):")
print(X_normalized[:5])

First 5 rows of the dataset:
  TL TM TR ML MM MR BL BM BR  class
0  x  x  x  x  o  o  x  o  o   True
1  x  x  x  x  o  o  o  x  o   True
2  x  x  x  x  o  o  o  o  x   True
3  x  x  x  x  o  o  o  b  b   True
4  x  x  x  x  o  o  b  o  b   True

Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 958 entries, 0 to 957
Data columns (total 10 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   TL      958 non-null    object
 1   TM      958 non-null    object
 2   TR      958 non-null    object
 3   ML      958 non-null    object
 4   MM      958 non-null    object
 5   MR      958 non-null    object
 6   BL      958 non-null    object
 7   BM      958 non-null    object
 8   BR      958 non-null    object
 9   class   958 non-null    bool  
dtypes: bool(1), object(9)
memory usage: 68.4+ KB
None

Unique values in each column:
TL: ['x' 'o' 'b']
TM: ['x' 'o' 'b']
TR: ['x' 'o' 'b']
ML: ['x' 'o' 'b']
MM: ['o' 'b' 'x']
MR: ['o' 'b' 'x']
BL

## Step 2: Build Neural Network

To build the neural network, you can refer to your own codes you wrote while following the [Deep Learning with Python, TensorFlow, and Keras tutorial](https://www.youtube.com/watch?v=wQ8BIBpya2k) in the lesson. It's pretty similar to what you will be doing in this lab.

1. Split the training and test data.
1. Create a `Sequential` model.
1. Add several layers to your model. Make sure you use ReLU as the activation function for the middle layers. Use Softmax for the output layer because each output has a single lable and all the label probabilities add up to 1.
1. Compile the model using `adam` as the optimizer and `sparse_categorical_crossentropy` as the loss function. For metrics, use `accuracy` for now.
1. Fit the training data.
1. Evaluate your neural network model with the test data.
1. Save your model as `tic-tac-toe.model`.

In [3]:
# your code here
# your code here
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from sklearn.model_selection import train_test_split

# Step 1: Split the training and test data
X_train, X_test, y_train, y_test = train_test_split(X_normalized, y, test_size=0.2, random_state=42)

# Step 2: Create a Sequential model
model = Sequential()

# Step 3: Add layers
# Input layer and hidden layers with ReLU activation
model.add(Dense(64, input_dim=X_train.shape[1], activation='relu'))  # First hidden layer
model.add(Dense(32, activation='relu'))  # Second hidden layer
model.add(Dense(16, activation='relu'))  # Third hidden layer

# Output layer with Softmax activation
model.add(Dense(2, activation='softmax'))  # Two output nodes (Player X wins: 1, doesn't win: 0)

# Step 4: Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# Step 5: Fit the training data
history = model.fit(X_train, y_train, epochs=20, batch_size=32, validation_split=0.2)

# Step 6: Evaluate the model on the test data
test_loss, test_accuracy = model.evaluate(X_test, y_test)
print(f"Test Accuracy: {test_accuracy:.2f}")

# Step 7: Save the model
# Save the model with the recommended .keras extension
model.save('tic-tac-toe.keras')


Epoch 1/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 1s 12ms/step - accuracy: 0.6224 - loss: 0.6551 - val_accuracy: 0.7013 - val_loss: 0.5975
Epoch 2/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6867 - loss: 0.6003 - val_accuracy: 0.7338 - val_loss: 0.5865
Epoch 3/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7039 - loss: 0.5763 - val_accuracy: 0.7078 - val_loss: 0.5811
Epoch 4/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7427 - loss: 0.5336 - val_accuracy: 0.7078 - val_loss: 0.5689
Epoch 5/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7517 - loss: 0.5340 - val_accuracy: 0.7468 - val_loss: 0.5587
Epoch 6/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7582 - loss: 0.5157 - val_accuracy: 0.7468 - val_loss: 0.5508
Epoch 7/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7558 - loss: 0.5203 - val_accuracy: 0.7468 - val_loss: 0.5308
Epoch 8/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7766 - loss: 0.4713 - val_accuracy: 0.7468 - val_loss

## Step 3: Make Predictions

Now load your saved model and use it to make predictions on a few random rows in the test dataset. Check if the predictions are correct.

In [4]:
# your code here
import numpy as np
import random
from tensorflow.keras.models import load_model

# Step 1: Load the saved model
model = load_model('tic-tac-toe.keras')  # Use the file name you saved earlier

# Step 2: Select random samples from the test dataset
num_samples = 5  # Number of random samples to check
random_indices = random.sample(range(len(X_test)), num_samples)
X_random = X_test[random_indices]
y_random = y_test.iloc[random_indices]

# Step 3: Make predictions
predictions = model.predict(X_random)

# Step 4: Interpret predictions
predicted_classes = np.argmax(predictions, axis=1)

# Step 5: Compare predictions with actual labels
for i, index in enumerate(random_indices):
    print(f"Sample {index}:")
    print(f"Predicted: {predicted_classes[i]} (Probability: {predictions[i][predicted_classes[i]]:.4f})")
    print(f"Actual: {y_random.iloc[i]}")
    print()


1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 58ms/step
Sample 42:
Predicted: 0 (Probability: 0.6917)
Actual: 0

Sample 166:
Predicted: 1 (Probability: 0.9547)
Actual: 1

Sample 22:
Predicted: 1 (Probability: 0.6915)
Actual: 0

Sample 48:
Predicted: 1 (Probability: 0.7787)
Actual: 0

Sample 191:
Predicted: 1 (Probability: 0.5930)
Actual: 0



## Step 4: Improve Your Model

Did your model achieve low loss (<0.1) and high accuracy (>0.95)? If not, try to improve your model.

But how? There are so many things you can play with in Tensorflow and in the next challenge you'll learn about these things. But in this challenge, let's just do a few things to see if they will help.

* Add more layers to your model. If the data are complex you need more layers. But don't use more layers than you need. If adding more layers does not improve the model performance you don't need additional layers.
* Adjust the learning rate when you compile the model. This means you will create a custom `tf.keras.optimizers.Adam` instance where you specify the learning rate you want. Then pass the instance to `model.compile` as the optimizer.
    * `tf.keras.optimizers.Adam` [reference](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Adam).
    * Don't worry if you don't understand what the learning rate does. You'll learn about it in the next challenge.
* Adjust the number of epochs when you fit the training data to the model. Your model performance continues to improve as you train more epochs. But eventually it will reach the ceiling and the performance will stay the same.

In [5]:
# your code here
# Rebuild the model with additional layers
model = Sequential()

# Input layer and hidden layers with ReLU activation
model.add(Dense(64, input_dim=X_train.shape[1], activation='relu'))  # First hidden layer
model.add(Dense(64, activation='relu'))  # Second hidden layer
model.add(Dense(32, activation='relu'))  # Third hidden layer
model.add(Dense(16, activation='relu'))  # Fourth hidden layer

# Output layer with Softmax activation
model.add(Dense(2, activation='softmax'))

# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])


In [6]:
from tensorflow.keras.optimizers import Adam

# Create a custom Adam optimizer with a lower learning rate
custom_optimizer = Adam(learning_rate=0.001)

# Compile the model with the custom optimizer
model.compile(optimizer=custom_optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])


In [7]:
# Train the model with more epochs
history = model.fit(X_train, y_train, epochs=50, batch_size=32, validation_split=0.2)


Epoch 1/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 2s 19ms/step - accuracy: 0.5042 - loss: 0.6997 - val_accuracy: 0.6818 - val_loss: 0.6295
Epoch 2/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.6549 - loss: 0.6237 - val_accuracy: 0.7013 - val_loss: 0.6033
Epoch 3/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.6500 - loss: 0.5936 - val_accuracy: 0.7403 - val_loss: 0.5887
Epoch 4/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7504 - loss: 0.5217 - val_accuracy: 0.7532 - val_loss: 0.5793
Epoch 5/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7831 - loss: 0.4937 - val_accuracy: 0.7597 - val_loss: 0.5443
Epoch 6/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.8135 - loss: 0.4629 - val_accuracy: 0.7662 - val_loss: 0.5309
Epoch 7/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.8207 - loss: 0.4232 - val_accuracy: 0.7662 - val_loss: 0.4938
Epoch 8/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8509 - loss: 0.4007 - val_accuracy: 0.7662 - val_loss

In [8]:
# Evaluate the model on the test data
test_loss, test_accuracy = model.evaluate(X_test, y_test)
print(f"Improved Test Accuracy: {test_accuracy:.2f}")


6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9263 - loss: 0.2472 
Improved Test Accuracy: 0.91


### ** **texto em negrito**Which approach(es) did you find helpful to improve your model performance?** **texto em negrito**

I found that increasing the number of epochs was very helpful in improving the model's performance. By training the model for 50 epochs instead of 20, it had more time to learn and optimize, which resulted in a higher test accuracy of 91%. Adding more layers also made a significant difference. The additional hidden layers allowed the model to better capture complex patterns in the data, leading to improved performance on both the validation and test sets. Adjusting the learning rate was another factor that contributed to the improvement. By using a custom Adam optimizer with a slightly lower learning rate, the training process became more stable, and the model was able to converge more effectively. Overall, these changes—more layers, a lower learning rate, and additional epochs—helped the model generalize better and achieve a noticeable improvement in accuracy.